### ISSUE TITLE
Server Side Rendering

### ISSUE THREAD

**Issue by geelen** [styled-components/styled-components#124]
Thought I'd open an issue to kick off the discussion. It's already possible because of the way we build on top of Glamor, but we haven't exposed it as an API. But basically, if you did this:

``` js
import styleSheet from 'styled-components/lib/models/StyleSheet'

/* before each render */
styleSheet.flush()

/* after each render */
styleSheet.rules().map(rule => rule.cssText).join('\n')
```

Then you should get the chunk of CSS you need for each request. Can someone with a server-rendered setup take a look and confirm this works, and maybe show how they'd be invoking it? I'd be happy enough to export something like:

``` js
import { serverStylesheet } from 'styled-components'

/* before each render */
serverStylesheet.reset()

/* after each render */
serverStylesheet.getCSS()
```

What do people think?

**Comment by pheuter** [comment]
I'll gladly help test this out! We currently avoid using styled components for most styles because of the unclear path to support SSR, would be great to move the inline styles into styled components if we can generate the CSS on the server.

Here is what I have setup so far:
```js
import styleSheet from 'styled-components/lib/models/StyleSheet';

styleSheet.flush();

let markup = renderToString(
  <Root store={store} Router={ServerRouter} routerProps={{ location, context }} />
);

console.log(styleSheet.rules());
```

Unfortunately, I'm getting back an empty array `[]`. I can confirm that one of the components nested inside `Root` is wrapped by `styled` and generates a className when the client loads.

**Comment by diegohaz** [comment]
I'm currently implementing SSR on a project and I managed to make this work without `styleSheet.flush()` (if I call this, the result is the same as @pheuter), only the styles from `injectGlobal` weren't included.

Once I finish it I will post the complete code and a demo here. (I'm trying to make the app work completely without javascript enabled on client).

**Comment by mxstbr** [comment]
> only the styles from injectGlobal weren't included.

Huh, that's interesting. I think I might accidentally use another `styleSheet` for that? /cc @geelen

**Comment by diegohaz** [comment]
Demo: https://arc.diegohaz.com (disable javascript, even the form works).
Code: https://github.com/diegohaz/arc/blob/universal-redux/src/server.js#L74

It doesn't work with `injectGlobal` (I removed).

The `serverStylesheet.getCSS()` API would be nice. 👍

**Comment by degroote22** [comment]
I noticed something.
I'm using Grid from 'grid-styled' and using `styleSheet.rules().map(rule => rule.cssText).join('\n')` seems to go just one level deep. It never rendered code of Grid, but rendered my code that used media queries.

code is like this:
```
const Wrapper = styled(Grid)`
  position: fixed;
  display: flex;
  justify-content: center;
  align-items: center;
  height: 40px;
  z-index: 0;
  top: 90px;
  z-index: 1;
  width: 100%;
  left: 0px;
  @media screen and (min-width: 52em){
    left: 16.66%;
  }
  @media screen and (min-width: 64em){
    left: 25%;
  }
`
```

Invoked as `<Wrapper xs={1} sm={1} md={2/3} lg={1/2}>{children}</Wrapper>`

gets rendered as

```
.iMsaJZ { 
  position: fixed;
  display: -webkit-box;display:flex;display:-webkit-flex;display:-ms-flexbox;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -moz-justify-content: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  -webkit-align-items: center;
  align-items: center;
  height: 40px;
  z-index: 0;
  top: 90px;
  z-index: 1;
  width: 100%;
  left: 0px;
 }
@media screen and (min-width: 52em){
  .iMsaJZ { 
    left: 16.66%;
  }
}
@media screen and (min-width: 64em){
  .iMsaJZ { 
    left: 25%;
  }
}

```

The element receives the correct classname on SSR, wich are two classes, one for my code and one for Grid's.
But there is no mention to the classname designed to Grid in the initial server response.
When I turn JS on then Grid's code appears to the browser.

I hope I made myself clear :)

**Comment by gbozee** [comment]
Based on @pheuter test, I get the same result,(the class is injected in the string markup) but not calling `stylesheet.flush()` before calling `styleSheet.rules()`  produces `TypeError: Cannot read property 'cssRules' of undefined`

**Comment by aesopwolf** [comment]
Ah, this is so great. I was able to get it working with [gatsby](https://github.com/gatsbyjs/gatsby) in a no time

```js
  render () {
    const styles = styleSheet.rules().map(rule => rule.cssText).join('\n');

    return (
      <html lang="en">
        <head>
          <style>
            {styles}
          </style>
        </head>
        <body>
          <div id="react-mount" dangerouslySetInnerHTML={{ __html: this.props.body }} />
          <script src={prefixLink(`/bundle.js?t=${BUILD_TIME}`)} />
        </body>
      </html>
    )
  }
```

I'm all for exposing a simpler api via `serverStylesheet.getCSS()`

**Comment by philpl** [comment]
@gbozee I ran into that problem. Turns out when (f.e.) webpack 2 is used, it loads the es bundle, due to the `package.json` js:next/module fields.

Until there is an API for the stylesheet that we can use, you should work around it using webpack's `resolve.alias` to load the `lib` folder instead.

**Comment by diegohaz** [comment]
> I think I might accidentally use another styleSheet for that?

@mxstbr I'm interested in solving this. Can you give me a way?

**Comment by MoOx** [comment]
Please get in touch when somebody start to handle this. I would like to get this to work out of the box in Phenomic (I just integrated Glamor and Aphrodite rendering and would like to cover all populars CSS in JS solutions).
I was thinking glamor integration will be enough but @philpl [just told me](https://github.com/MoOx/phenomic/commit/6d703cb1894f8cfdeb026a2d903448541a0f6ea5#commitcomment-19860196) it won't work :(

**Comment by thisguychris** [comment]
@aesopwolf did you get the same problem as @diegohaz of not having styles included from injectGlobal calls?

I think its a general consensus that this is working out of the box.  The only minor caveat is styles from injectGlobal isn't included.  But fixing that is trivial I guess?

**Comment by aesopwolf** [comment]
@thisguychris I just ran a quick test with `injectGlobal` and I didn't get the same problem. My document head included a single style tag with everything expected.

Input:

```js
const Title = styled.h1`
  font-size: 1.5em;
  text-align: center;
  color: palevioletred;
`;

injectGlobal`
  body {
    background: red;
  }
`;
```

Output:
```html
<style>
    .kZQsGZ {
        font-size: 1.5em;
        text-align: center;
        color: palevioletred;
    }
    body {
        background: red;
    }
</style>
```

**Comment by geelen** [comment]
Yeah the implementation for `injectGlobal` uses the same instance of `StyleSheet` under the hood so what @aesopwolf is getting looks right..

**Comment by thisguychris** [comment]
@geelen so do we get a green light for the exposed api? aka serverStylesheet.getCSS()?

Also, @aesopwolf would you mind giving me a test repo of your SSR to test my PR with?

**Comment by mxstbr** [comment]
I would love an exposed API, would love for you to look into that @thisguychris!

**Comment by diegohaz** [comment]
You guys are right, `injectGlobal` is working. It was a problem in my code.
I was using it inside `componentDidMount` (which doesn't run in ssr).

Sorry for the mess.

**Comment by diegohaz** [comment]
Now this repo is working with `styled-components`, `injectGlobal`, SSR and everything: https://github.com/diegohaz/arc/tree/fullstack

@thisguychris

**Comment by tomazy** [comment]
It would be great if `.flush()` worked. The problem may be somewhere here:
https://github.com/styled-components/styled-components/blob/master/src/vendor/glamor/sheet.js#L88
```js
     this.sheet  = {
        cssRules: [],
        insertRule: rule => {
          const serverRule = { cssText: rule }
          this.sheet.cssRules.push(serverRule)              
          return {serverRule, appendRule: (newCss => serverRule.cssText += newCss)}  
        }
      }
```

`flush()` on server does this:
```js
this.sheet.cssRules = []
```

So after flushing `appendRule` will be appending text to a dangling `serverRule` that is not in `this.sheet.cssRules` anymore.

**Comment by thisguychris** [comment]
@diegohaz tried fullstack branch getting this error
`[0] MongoDB connection error: MongoError: failed to connect to server [0.0.0.0:27017] on first connect`

**Comment by diegohaz** [comment]
@thisguychris It's because that branch requires the MongoDB service to run. You can use the `universal-redux` branch as well (it doesn't require mongod).

**Comment by thisguychris** [comment]
@diegohaz looks like its working 👍 
http://take.ms/r9QiC

So we don't need flush() / reset() for SSR?

**Comment by tomazy** [comment]
Currently styled components will inject the styles even though they are rendered on the server and there is no need for that. Are you guys planning to do something about it? Ideal solution would be to reuse the prerendered styles and inject only these that are necessary. 

Maybe it would be possible to render the generated style tag(s) within \<body/\> with React? Then React would update their contents only when the change.

**Comment by jorilallo** [comment]
Being relatively new to `styled-components` I started adopting SSR for a hack project. For some reason I'm getting `TypeError: Cannot set property 'cssRules' of undefined` with or without `.flush()`. Execution will actually fail to this error already when calling `flush()` prior to `renderToString`.

Here's what logging out `styleSheet` yields:

```
e {
  isSpeedy: false,
  sheet: undefined,
  tags: [],
  maxLength: 40,
  ctr: 0 }
```

and here's my sample code for sanity check:

```js
import express from 'express';
import React from 'react';
import { renderToString } from 'react-dom/server';
import fs from 'fs';
import App from './App.js';
import template from './template.js';
import styleSheet from 'styled-components/lib/models/StyleSheet';

const config = JSON.parse(fs.readFileSync('./build/asset-manifest.json', { encoding: 'utf8' }));

const jsPath = config ? config[Object.keys(config).find(fileName => fileName.endsWith('.js'))] : '/bundle.js';
const cssPath = config ? config[Object.keys(config).find(fileName => fileName.endsWith('.css'))] : null;

styleSheet.flush();
const markup = renderToString(<App />);
const styles = styleSheet.rules().map(rule => rule.cssText).join('\n');

const server = express();
server.use('/static', express.static('build/static'));
server.get('*', (req, res) => res
  .send(template(markup, jsPath, cssPath, styles)));

const port = process.env.PORT || 3000;
server.listen(port);
``` 

SSR is working fine aside from styled components.

**Comment by dignifiedquire** [comment]
I am trying to use hedron, but it seems the styles from it are not present. But my custom styles are rendered fine. Any ideas why this is happening/how to work around?

**Comment by diegohaz** [comment]
Don't know if it's better to create a new issue, but I'm seeing this kind of warning on server when passing custom props to a styled component:
```
Warning: Unknown props `transparent`, `light` on <a> tag. Remove these props from the element. For details, see https://fb.me/react-unknown-prop
```

It happens only on server (terminal), not on browser.

**Comment by clempat** [comment]
@jorilallo I had quite similar issue. In my case I solved it by adding 'styled-components/lib/models/StyleSheet' in externals in webpack 2.

externals: {
  'styled-components/lib/models/StyleSheet': 'commonjs styled-components/lib/models/StyleSheet',
}

it was not there because I was adding only the node_modules root folder to externals.

**Comment by code-punkt** [comment]
I'm not sure why this is so hard using `styled-components`. It's already possible with glamor, which it seems this is based on, right? Can't we just simply use the underlying glamor API or parts of it and re-expose?

**Comment by thisguychris** [comment]
@code-punkt see #214 :)

**Comment by rauchg** [comment]
Curious about why both `.reset` and `.flush` are required? Wouldn't the `flush` after the render call be sufficient, since it returns the styles and resets the state for the next call?

**Comment by SachaG** [comment]
@aesopwolf thanks for the code snippet, worked with Gatsby for me too. But I think it should be `<style dangerouslySetInnerHTML={{ __html: styles }} />`, otherwise special characters will get escaped out.

**Comment by eXon** [comment]
@tomazy You can always remove your SSR style block once the client-side rendering is done. That way, your page will look great while the app is loading and you will not get duplicate styles.

```js
ReactDOM.render(<App />, document.getElementById('app'));
document.getElementById('ssr-style').remove();
```

**Comment by codepunkt** [comment]
@exOn yes, but styled-components could also just stop injecting it on the client-side again so you don't have to hack stuff like that together ;-)

See #214 for more discussion about this.

**Comment by teonik** [comment]
@tomazy I used something similar to remove the server rendered critical-css block when react and styled-components kicked in on the client side.
The problem with this approach is that it will also remove any styles you injected with injectGlobals,Those styles, if rendered on the server, will end up in the ssr block.
It is also feels like a dirty hack compared to the rest of the experience offered by styled-components 😛 

Of course there are simpler ways to stick some global css in the pre-rendered page's `head` section but the point is that it would be nice to have proper support for it.

**Comment by kristojorg** [comment]
PSA: I was seeing `TypeError: Cannot read property 'cssRules' of undefined` when using this line of code:

`const styles = styleSheet.rules().map(rule => rule.cssText).join('\n');`

*The fix:* If you are seeing this as well, make sure you are rendering at least one styled component in your app. 

For me, I was just installing `styled-components` for the first time, and hadn't used one yet, and was therefore seeing this error.

**Comment by kristojorg** [comment]
I have submitted a PR to fix this issue [here](https://github.com/styled-components/styled-components/pull/391). Hope this helps some!

**Comment by vernondegoede** [comment]
I'm using 'hedron' (which uses styled-components) as well and my stylesheet rules do get mapped. However, I would like the 'hedron' stylesheet rules to get mapped as well. At the moment this doesn't seem to be happening, since the following error is being thrown:

```
Warning: React attempted to reuse markup in a container but the checksum was invalid. This generally means that you are using server rendering and the markup generated on the server was not what the client was expecting. React injected new markup to compensate which works but you have lost many of the benefits of server rendering. Instead, figure out why the markup being generated is different on the client or server:
```

What would be the best way to solve this issue? Should I submit a PR for 'hedron' and add a method that returns `styleSheet.rules().map(rule => rule.cssText).join("\n");`? And invoke hedron's method in my page head, where I run this method for my own styleSheet rules as well?

**Comment by karthikiyengar** [comment]
> I am trying to use hedron, but it seems the styles from it are not present. But my custom styles are rendered fine. Any ideas why this is happening/how to work around?

@dignifiedquire - Could you please tell me how you worked around this? I'm having the exact same issue with hedron.
@diegohaz - Using the universal-redux branch for this, is there anything that I should look into particularly?

**Comment by kristojorg** [comment]
@karthikiyengar Hedron is using styled-components which I believe means you need to import both your own server markup and theirs.

**Comment by karthikiyengar** [comment]
@kristojorg - Thanks for your answer.

Please disregard, my vendor assets weren't being generated properly, leading to the issue.

**Comment by mxstbr** [comment]
This discussion has now been moved to #386.
